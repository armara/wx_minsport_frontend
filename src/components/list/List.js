import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Image from 'components/image/Image'

import 'components/list/List.scss'

const rowRenderer = item => {
  if (item.type === 'hr') {
    return <hr />
  }

  const { spanClassName } = item

  return (
    <>
      <div>
        {item.image && <Image src={item.image} />}
        <h4>{item.field}</h4>
      </div>

      {!item.value ? (
        '---'
      ) : (
        <span className={spanClassName}>{item.value}</span>
      )}
    </>
  )
}

const List = ({
  id,
  keyId,
  items,
  children,
  className,
  rowClassName,
  getRowClassName,
  notFoundItem,
  childrenProps,
}) => (
  <ul className={classnames([className, 'base-list'])} id={id}>
    {items.length === 0 ? (
      <li className="base-list_not-found">{notFoundItem}</li>
    ) : (
      items.map((item, i) => (
        <li
          key={item[keyId]}
          className={classnames([rowClassName, getRowClassName(item, i)])}
        >
          {children ? children(item, childrenProps) : rowRenderer(item)}
        </li>
      ))
    )}
  </ul>
)

List.defaultProps = {
  id: null,
  keyId: 'key',
  className: '',
  rowClassName: '',
  children: '',
  getRowClassName: () => {},
  childrenProps: {},
  notFoundItem: 'Не найдено',
}

List.propTypes = {
  id: PropTypes.string,
  keyId: PropTypes.string,
  className: PropTypes.string,
  rowClassName: PropTypes.string,
  getRowClassName: PropTypes.func,
  notFoundItem: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      image: PropTypes.string,
      field: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      type: PropTypes.oneOf(['hr']),
    })
  ).isRequired,
  childrenProps: PropTypes.shape({}),
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
}

export default List
