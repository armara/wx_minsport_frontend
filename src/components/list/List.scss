@import "styles/mixins";

.base-list {
  .base-list_not-found {
    font-style: italic;
  }

  hr {
    width: 100%;
    margin-top: 15px;
    margin-bottom: 15px;
  }

  li + li {
    margin-top: 15px;
  }

  li {
    display: flex;
    justify-content: space-between;

    h4 {
      @include roboto-regular-text($small, 18px, $evil-gray);
      display: flex;
      align-items: center;
    }

    & > div {
      display: flex;

      img {
        margin-right: 10px;
      }
    }

    & > span {
      @include nav-header-text($small, '18px', $black);
      display: flex;
      align-items: center;
    }
  }
}

.tabs-list {
  padding-top: 0;

  li {
    height: 45px;
  }

  li + li {
    margin-top: 0;
  }
}