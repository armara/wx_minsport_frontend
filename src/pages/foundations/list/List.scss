@import "styles/mixins";

.foundations-list {
  height: 100%;
  display: flex;
  flex-flow: column;

  .foundations-list_filters {
    display: flex;
    align-items: center;
    padding-top: 25px;
    padding-bottom: 25px;
  }

  .foundations-list_filters_dropdown + .foundations-list_filters_dropdown {
    margin-left: 11px;
  }

  .foundations-list_filters_types-trigger,
  .foundations-list_filters_types-menu {
    min-width: 220px;
  }

  .foundations-list_filters_types-trigger,
  .foundations-list_filters_trigger {
    button.is-white {
      @include nav-header-text($small, '18px', $evil-gray);
      width: 100%;
    }
  }

  .foundations-list_filters_trigger,
  .foundations-list_filters_menu {
    min-width: 200px;
  }

  .foundations-list_search-input {
    margin-left: auto;
  }

  .foundations-list_table {
    flex-grow: 1;
    align-items: flex-start;
    margin-bottom: 20px;
  }
}