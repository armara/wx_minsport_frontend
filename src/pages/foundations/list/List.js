import React, { useState, useEffect, useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import debounce from 'lodash.debounce'

import TableCellActions from 'components/table/tableCellActions/TableCellActions'
import TableArrowPaginator from 'components/table/tableArrowPaginator/TableArrowPaginator'
import TableCellFoundation from 'components/table/tableCellIcon/TableCellIcon'
import TableCellText from 'components/table/tableCellText/TableCellText'
import Table from 'components/table/Table'
import AddFoundationButton from 'pages/foundations/components/addFoundationButton/AddFoundationButton'
import FilterPanel from 'components/filterPanel/FilterPanel'

import history from 'utils/utils'
import { facilitiesPrefix as prefix } from 'utils/auth'
import {
  setModal,
  setCurrentContract,
} from 'store/actions/foundation/foundation'
import {
  getAllContracts,
  getContracts,
  searchByName,
  filterBySport,
  filterByType,
} from 'store/actionCreators/foundation/foundation'
import {
  foundationAllowedSportsSelector,
  foundationAllowedTypesSelector,
  foundationCurrentContractsSelector,
  foundationCurrentPageSelector,
  foundationLoadingSelector,
  foundationPageSizeSelector,
  foundationPagesSelector,
  foundationSportFilterSelector,
  foundationTotalSelector,
  foundationTypeFilterSelector,
} from 'store/selectors/foundation'

import './List.scss'

const columnsCallback = (onEditClick, onDeleteClick) => [
  {
    id: 'Юридическое название договора',
    Header: 'Юридическое название договора',
    isFilterable: true,
    accessor: row => <TableCellFoundation foundation={row.legalData} />,
  },
  {
    id: 'Тип договора',
    Header: 'Тип договора',
    isFilterable: true,
    accessor: row => <TableCellText text={row.contractTitle} />,
  },
  {
    id: 'Заказчик',
    Header: 'Заказчик',
    isFilterable: true,
    accessor: row => <TableCellText text={row.organizationName} />,
  },
  {
    id: 'tin',
    Header: 'ИНН',
    isFilterable: true,
    accessor: row => <TableCellText text={row.organizationTin} />,
  },
  {
    id: 'Вид спорта',
    Header: 'Вид спорта',
    isFilterable: true,
    accessor: row => <TableCellText text={row.sportTitle} />,
  },
  {
    id: 'Регистрация',
    Header: 'Регистрация',
    isFilterable: true,
    accessor: row => <TableCellText text={row.reg} />,
  },
  {
    id: 'Действия',
    Header: ' ',
    width: 80,
    isFilterable: false,
    accessor: row => (
      <TableCellActions
        onEditClick={event => onEditClick(event, row.id)}
        onDeleteClick={event => onDeleteClick(event, row.id)}
      />
    ),
  },
]

function List({
  setModalState,
  setContract,
  currentPage,
  pageSize,
  pages,
  filterTypeValue,
  contracts,
  total,
  sportFilterValue,
  fetchContracts,
  loading,
  allowedSports,
  allowedTypes,
  handleSportChange,
  handleTypeChange,
  handleNameChange,
  fetchPage,
}) {
  const [searchText, setSearchText] = useState('')

  const onEditClick = useCallback((event, id) => {
    event.stopPropagation()
    history.push(`/${prefix}/foundations/edit/${id}`)
  }, [])

  const onDeleteClick = useCallback(
    (event, id) => {
      event.stopPropagation()
      setContract({ id })
      setModalState(true)
    },
    [setContract, setModalState]
  )

  const columns = columnsCallback(onEditClick, onDeleteClick)

  useEffect(() => {
    fetchContracts()
  }, [fetchContracts])

  const handleOnBackClick = () => {
    if (currentPage !== 0) {
      fetchPage(currentPage - 1)
    }
  }

  const handleOnForwardClick = () => {
    if (currentPage !== pages) {
      fetchPage(currentPage + 1)
    }
  }

  const debouncedOnChange = useCallback(
    debounce(value => handleNameChange(value), 200),
    []
  )

  const onChange = event => {
    setSearchText(event.target.value)
    debouncedOnChange(event.target.value)
  }

  return (
    <div className="foundations-list">
      <FilterPanel
        onSearch={onChange}
        filters={[
          {
            id: 1,
            items: allowedTypes,
            selectedItem: filterTypeValue,
            onSelect: handleTypeChange,
            inputPlaceholder: 'Выбрать тип организации',
          },
          {
            id: 2,
            items: allowedSports,
            onSelect: handleSportChange,
            inputPlaceholder: 'Выбрать вид спорта',
            selectedItem: sportFilterValue,
          },
        ]}
        searchPlaceholder="Поиск договоров"
        searchValue={searchText}
      />

      <Table
        columns={columns}
        data={contracts}
        loading={loading}
        onRowClick={rowInfo =>
          history.push(`/${prefix}/foundations/${rowInfo.original.id}`)
        }
      />

      <TableArrowPaginator
        totalRows={total}
        currentPage={currentPage}
        onBackClick={handleOnBackClick}
        onForwardClick={handleOnForwardClick}
        pageSize={pageSize}
        pages={pages}
      >
        <AddFoundationButton />
      </TableArrowPaginator>
    </div>
  )
}

List.propTypes = {
  contracts: PropTypes.arrayOf(PropTypes.object).isRequired,
  fetchPage: PropTypes.func.isRequired,
  fetchContracts: PropTypes.func.isRequired,
  handleSportChange: PropTypes.func.isRequired,
  handleTypeChange: PropTypes.func.isRequired,
  handleNameChange: PropTypes.func.isRequired,
  total: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  pages: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  loading: PropTypes.bool.isRequired,
  filterTypeValue: PropTypes.shape({}).isRequired,
  sportFilterValue: PropTypes.shape({}).isRequired,
  allowedSports: PropTypes.arrayOf(PropTypes.object).isRequired,
  allowedTypes: PropTypes.arrayOf(PropTypes.object).isRequired,
  setContract: PropTypes.func.isRequired,
  setModalState: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  contracts: foundationCurrentContractsSelector(state),
  total: foundationTotalSelector(state),
  pageSize: foundationPageSizeSelector(state),
  pages: foundationPagesSelector(state),
  currentPage: foundationCurrentPageSelector(state),
  loading: foundationLoadingSelector(state),
  allowedSports: foundationAllowedSportsSelector(state),
  allowedTypes: foundationAllowedTypesSelector(state),
  filterTypeValue: foundationTypeFilterSelector(state),
  sportFilterValue: foundationSportFilterSelector(state),
})

const mapDispatchToProps = dispatch => ({
  setContract: id => dispatch(setCurrentContract(id)),
  setModalState: isModalOpen => dispatch(setModal(isModalOpen)),
  fetchPage: page => dispatch(getContracts(page)),
  fetchContracts: () => dispatch(getAllContracts()),
  handleSportChange: item => dispatch(filterBySport(item)),
  handleTypeChange: item => dispatch(filterByType(item)),
  handleNameChange: item => dispatch(searchByName(item)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(List)
