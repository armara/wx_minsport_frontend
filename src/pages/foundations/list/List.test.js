import React from 'react'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'

import List from 'pages/foundations/list/List'
import initialState from 'tests/initialStoreMock'

configure({ adapter: new Adapter() })
const mockStore = configureStore()

describe('<List />', () => {
  let store

  beforeEach(() => {
    // creates the store with any initial state or middleware needed
    store = mockStore(initialState)
    localStorage.setItem('user', '{"test":"ok"}')
  })

  it('renders correctly default props', () => {
    const tree = renderer
      .create(
        <Router initialEntries={['/foundations']} initialIndex={0}>
          <Provider store={store}>
            <List />
          </Provider>
        </Router>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
