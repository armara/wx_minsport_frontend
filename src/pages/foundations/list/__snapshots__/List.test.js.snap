// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> renders correctly default props 1`] = `
<div
  className="foundations-list"
>
  <div
    className="list-filters"
  >
    <div
      className="search-input control has-icons-left filters-search"
    >
      <input
        className="input is-small base-input input"
        disabled={false}
        onChange={[Function]}
        placeholder="Поиск договоров"
        type="text"
        value=""
      />
      <span
        className="icon is-small is-left"
      >
        <i
          className="ic-search"
        />
      </span>
    </div>
    <div
      className="dropdown"
      id="filters-1"
    >
      <div
        className="dropdown-trigger"
      >
        <p
          className="control has-icons-right has-icons-left input-trigger"
        >
          <input
            className="is-small base-input input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Выбрать тип организации"
            type="text"
            value="Все типы организаций"
          />
          <span
            className="icon is-small is-left"
          >
            <i
              className="ic-place"
            />
          </span>
          <span
            className="icon is-small is-right"
          >
            <i
              className="ic-keyboard-arrow-down"
            />
          </span>
        </p>
      </div>
      <div
        className="dropdown-menu"
        role="menu"
      >
        <ul
          className="dropdown-content menu-list"
        >
          <button
            className="dropdown-item button base-button"
            disabled={false}
            id={null}
            onClick={[Function]}
            title=""
            type="button"
          >
            
            Все типы организаций
          </button>
        </ul>
      </div>
    </div>
    <div
      className="dropdown"
      id="filters-2"
    >
      <div
        className="dropdown-trigger"
      >
        <p
          className="control has-icons-right has-icons-left input-trigger"
        >
          <input
            className="is-small base-input input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Выбрать вид спорта"
            type="text"
            value="Все виды спорта"
          />
          <span
            className="icon is-small is-left"
          >
            <i
              className="ic-place"
            />
          </span>
          <span
            className="icon is-small is-right"
          >
            <i
              className="ic-keyboard-arrow-down"
            />
          </span>
        </p>
      </div>
      <div
        className="dropdown-menu"
        role="menu"
      >
        <ul
          className="dropdown-content menu-list"
        >
          <button
            className="dropdown-item button base-button"
            disabled={false}
            id={null}
            onClick={[Function]}
            title=""
            type="button"
          >
            
            Все виды спорта
          </button>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="table-component"
  >
    <div
      className="ReactTable"
      style={Object {}}
    >
      <div
        className="rt-table"
        role="grid"
      >
        <div
          className="rt-thead -header"
          style={
            Object {
              "minWidth": "680px",
            }
          }
        >
          <div
            className="rt-tr"
            role="row"
          >
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                Юридическое название договора
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                Тип договора
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                Заказчик
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                ИНН
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                Вид спорта
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "100 0 auto",
                  "maxWidth": null,
                  "width": "100px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                Регистрация
              </div>
            </div>
            <div
              className="rt-th"
              onClick={[Function]}
              role="columnheader"
              style={
                Object {
                  "flex": "80 0 auto",
                  "maxWidth": "80px",
                  "width": "80px",
                }
              }
              tabIndex="-1"
            >
              <div
                className=""
              >
                 
              </div>
            </div>
          </div>
        </div>
        <div
          className="rt-tbody"
          style={
            Object {
              "minWidth": "680px",
            }
          }
        />
      </div>
      <div
        className="rt-noData"
      >
        Нет данных
      </div>
      <div
        className="-loading"
      >
        <div
          className="-loading-inner"
        >
          Загрузка...
        </div>
      </div>
    </div>
  </div>
  <div
    className="table-arrow-paginator"
  >
    <div>
      <button
        className="add-foundation button base-button"
        disabled={false}
        id={null}
        onClick={[Function]}
        title=""
        type="button"
      >
        <span
          className="icon"
        >
          <i
            className="ic-add"
          />
        </span>
        <span>
          Добавить договор
        </span>
      </button>
    </div>
    <span>
      <span>
        1 - 0 из 0
      </span>
      <button
        className="is-small button base-button disabled"
        disabled={true}
        id={null}
        onClick={[Function]}
        title=""
        type="button"
      >
        <span
          className="icon is-small"
        >
          <i
            className="ic-arrow-back"
          />
        </span>
        
      </button>
      <button
        className="is-small button base-button"
        disabled={false}
        id={null}
        onClick={[Function]}
        title=""
        type="button"
      >
        <span
          className="icon is-small"
        >
          <i
            className="ic-arrow-forward"
          />
        </span>
        
      </button>
    </span>
  </div>
</div>
`;
